sympy/simplify/cse_main.py

```diff
@@ -212,57 +212,99 @@ def _find_opts(expr):
 
     ## Process Adds and commutative Muls
 
-    def _match_common_args(Func, funcs):
-        if order != 'none':
-            funcs = list(ordered(funcs))
-        else:
-            funcs = sorted(funcs, key=lambda x: len(x.args))
-
-        func_args = [set(e.args) for e in funcs]
-        for i in range(len(func_args)):
-            for j in range(i + 1, len(func_args)):
-                com_args = func_args[i].intersection(func_args[j])
-                if len(com_args) > 1:
-                    com_func = Func(*com_args)
-
-                    # for all sets, replace the common symbols by the function
-                    # over them, to allow recursive matches
-
-                    diff_i = func_args[i].difference(com_args)
-                    func_args[i] = diff_i | {com_func}
-                    if diff_i:
-                        opt_subs[funcs[i]] = Func(Func(*diff_i), com_func,
-                                                  evaluate=False)
+    # split muls into commutative/noncommutative
+    comutative_muls = set()
+    noncommutative_muls = set()
+    noncommutative_matmuls = set()
+    for m in muls:
+        c, nc = m.args_cnc(cset=True)
+        c_mul = m.func(*c)
+        nc_mul = m.func(*nc)
+        if nc:
+            if c_mul == 1:
+                new_obj = nc_mul
+            else:
+                new_obj = m.func(c_mul, nc_mul, evaluate=False)
+            if isinstance(nc_mul, Mul):
+                noncommutative_muls.add(nc_mul)
+            elif isinstance(nc_mul, MatMul):
+                noncommutative_matmuls.add(nc_mul)
+            opt_subs[m] = new_obj
+        if len(c) > 1:
+            comutative_muls.add(c_mul)
+
+    opt_subs.update(match_common_args(Add, adds, order=order))
+    opt_subs.update(match_common_args(Mul, comutative_muls, order=order))
+
+    opt_subs.update(match_common_args_nc(Mul, noncommutative_muls))
+    opt_subs.update(match_common_args_nc(MatMul, noncommutative_matmuls))
+
+    return opt_subs
+
+
+def match_common_args(Func, funcs, order='canonical'):
+    opt_subs = {}
 
-                    diff_j = func_args[j].difference(com_args)
-                    func_args[j] = diff_j | {com_func}
-                    opt_subs[funcs[j]] = Func(Func(*diff_j), com_func,
+    if order != 'none':
+        funcs = list(ordered(funcs))
+    else:
+        funcs = sorted(funcs, key=lambda x: len(x.args))
```

Comment:
I didn't write this code. I just moved the function definition out so that it was more accessible (I don't like functions defined within functions). If you add `?w=0` to the end of the PR url it will show you just the new stuff.